lua/lazyvim/plugins/extras/lang/python.lua

```diff
@@ -0,0 +1,116 @@
+return {
+  {
+    "nvim-treesitter/nvim-treesitter",
+    opts = function(_, opts)
+      if type(opts.ensure_installed) == "table" then
+        vim.list_extend(opts.ensure_installed, { "ninja", "python", "rst", "toml" })
+      end
+    end,
+  },
+  {
+    "jose-elias-alvarez/null-ls.nvim",
+    opts = function(_, opts)
+      local nls = require("null-ls")
+      opts.sources = opts.sources or {}
+      vim.list_extend(opts.sources, {
+        -- Order of formatters matters. They are used in order of appearance.
+        nls.builtins.formatting.ruff.with({ filetypes = { "python" }, prefer_local = ".venv/bin" }),
+        nls.builtins.formatting.black.with({ filetypes = { "python" }, prefer_local = ".venv/bin" }),
+      })
+    end,
+    dependencies = {
+      "mason.nvim",
+      opts = function(_, opts)
+        opts.ensure_installed = opts.ensure_installed or {}
+        vim.list_extend(opts.ensure_installed, { "black", "ruff" })
+      end,
+    },
+  },
+  {
+    "neovim/nvim-lspconfig",
+    opts = {
+      servers = {
+        pyright = {
+          settings = {
+            pyright = {
+              disableLanguageServices = false,
+              disableOrganizeImports = false,
+            },
+            python = {
+              analysis = {
+                autoImportCompletions = true,
+                autoSearchPaths = true,
+                diagnosticMode = "workspace",
+                typeCheckingMode = "basic",
+                useLibraryCodeForTypes = true,
+              },
+            },
+          },
+        },
+        ruff_lsp = {
```

Comment:
If it's added in null-ls, why add it here? (or vice versa)

Reply:
`ruff` is used here as diagnostics and formatting with ruff is done via null-ls

Reply:
null-ls can also use ruff for diagnostic, is there any difference in moving it there?

Reply:
> `ruff` is used here as diagnostics and formatting with ruff is done via null-ls

It's really `ruff_lsp` doing the diagnostics. They're separate.

Reply:
> null-ls can also use ruff for diagnostic, is there any difference in moving it there?

`null-ls` cannot use `ruff` as `code_actions`. `ruff_lsp` solves this issue and therefore `ruff_lsp` is used as diagnostics and code_actions server and formatting with `ruff` is done via `null-ls`